### PR TITLE
Support multiple-array UDFs in R

### PR DESCRIPTION
**Pull request by johnkerl**
# Context

TileDB-internal task 10789.

# Example

```
#!/usr/bin/env Rscript

library(tiledbcloud)
library(tiledb)

# Not doing tiledbcloud::login here; re-using cloud-login.py having written ~/.tiledb/cloud.json.

myfunc <- function(df1, df2) {
  vec1 <- as.vector(df1[["a"]])
  vec2 <- as.vector(df2[["a"]])
  list(
    len=length(vec1) + length(vec2),
    min=min(vec1) + min(vec2),
    med=median(vec1) + median(vec2),
    max=max(vec1) + max(vec2)
  )
}

details1 <- tiledbcloud::UDFArrayDetails$new(
  uri="tiledb://demo/quickstart_dense",
  ranges=QueryRanges$new(
    layout=Layout$new('row-major'),
    ranges=list(cbind(1,4),cbind(1,4))
  ),
  buffers=list("a")
)

details2 <- tiledbcloud::UDFArrayDetails$new(
  uri="tiledb://demo/quickstart_sparse",
  ranges=QueryRanges$new(
    layout=Layout$new('row-major'),
    ranges=list(cbind(1,2),cbind(1,4))
  ),
  buffers=list("a")
)

result <- tiledbcloud::execute_multi_array_udf(
  namespace="demo",
  array_list=list(details1, details2),
  udf=myfunc,
)
print(result)
```

# Documentation

On a separate PR, we should split apart https://github.com/TileDB-Inc/TileDB-Cloud-R/blob/master/howToBuild.md into a for-developers part on OpenAPI regeneration, and a new, separate, for-users page including the above example and more.

# Discussion

The syntax

```
details1 <- tiledbcloud::UDFArrayDetails$new(
  uri=uri1,
  ranges=QueryRanges$new(
    layout=Layout$new('row-major'),
    ranges=list(cbind(1,4),cbind(1,4))
  ),
  buffers=list("a")
)
```

might be user-friendlified a bit (let's discuss). Namely, replacing an explicit R6 class with a calling syntax like

```
details1 <- tiledbcloud::array_defaults(
  uri=uri1, 
  layout='row-major',
  ranges=list(cbind(1,4),cbind(1,4)),
  buffers=list("a")
)
```

We might:

* Do the last example (proposal) on this PR
* Do it on a separate PR
* Do something else
* Leave the calling syntax as-is with explicit `UDFArrayDetails$new`

### CODE REVIEW

File: R/udf_array_details.R [hunk 1]
```diff
@@ -54,13 +54,11 @@ UDFArrayDetails <- R6::R6Class(
       }
       if (!is.null(self$`ranges`)) {
         UDFArrayDetailsObject[['ranges']] <-
-          ####self$`ranges`$toJSON()
-          jsonlite::toJSON(as.integer(serialize(self$ranges, NULL)))
+          self$`ranges`$toJSON()
```
Comment (eddelbuettel) [inline R/udf_array_details.R:57]: This does not need backtick as `ranges` is a normal identifier string.

File: R/udf_array_details.R [hunk 2]
```diff
@@ -54,13 +54,11 @@ UDFArrayDetails <- R6::R6Class(
...
       }
       if (!is.null(self$`buffers`)) {
         UDFArrayDetailsObject[['buffers']] <-
-          ####self$`buffers`
-          jsonlite::toJSON(as.integer(serialize(self$buffers, NULL)))
+          self$`buffers`
```
Comment (eddelbuettel) [inline R/udf_array_details.R:61]: As above, no backticks needed.

File: man/execute_multi_array_udf.Rd [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+\arguments{
+\item{namespace}{Namespace within TileDB cloud.}
+
+\item{array_list}{Names of the arrays. E.g. if one URI is \code{tiledb://hello/earth}
+and another is \code{tiledb://hello/mars} ##' then the namespace is "hello" and the array_list is
+\code{list("earth", "mars")}.}
+
+\item{udf}{An R function which takes an array of dataframes as argument.}
+
+\item{queryRanges}{List of two-column matrices, one matrix per dimension,
```
Comment (eddelbuettel) [inline man/execute_multi_array_udf.Rd:24]: queryRanges is not an argument above, and the arguments layout and result_format seem undefined.

**Review by eddelbuettel — COMMENTED** [review]
Seems mostly fine but I left some minor comments.

File: R/udf_array_details.R (hunk 1 above)
Comment (johnkerl) [reply R/udf_array_details.R:57]: thanks @eddelbuettel ! I'll edit this & some in the vicinity but most of these are among the ~30KLOC generated by OpenAPI which backticks everything in an overabundance of caution. :(

File: R/udf_array_details.R (hunk 2 above)
Comment (johnkerl) [reply R/udf_array_details.R:61]: as above

File: R/udf_array_details.R (hunk 1 above)
Comment (eddelbuettel) [reply R/udf_array_details.R:57]: Oh, I missed that.  If it is robot-generated (which may unconditionally toss backticks in) then never mind.  Not worth chasing.

**Comment by johnkerl** [comment]
@eddelbuettel thanks! Ready for round two. :)

**Review by eddelbuettel — APPROVED** [review]
:+1:

File: R/udf_array_details.R (hunk 1 above)
Comment (johnkerl) [reply R/udf_array_details.R]: Euh, it was me as a non-human hand-editing what a robot had written. :D

**Pull request merged**